### PR TITLE
Qt refactoring: basic concept, LookupModule

### PR DESCRIPTION
**Pull request by Karry**
Hi all. 

This is first step of Qt API refactoring that I mention here: https://github.com/Framstag/libosmscout/issues/284#issuecomment-295604526

@Framstag @aquiles2k Please review API changes. Rendering is still in `DBThread`, I started with simple `LookupModule` that is used from `MapObjectInfoModel`. What was changed:

## Library initialisation

Instead of usage `DBThread` singleton as a central point where you have to choose between plane and tiled variant of rendering, I created `OSMScoutQt` singleton that is used to accessing `DBThread`, `Settings` and creating modules for models... Its initialisation uses builder pattern:

```
OSMScoutQt::NewInstance()
     .WithStyleSheetDirectory(stylesheetFile.dir().path())
     .WithStyleSheetFile(stylesheetFile.fileName())
     .WithIconDirectory(iconDirectory)
     .WithMapLookupDirectories(mapLookupDirectories)
     .WithOnlineTileProviders(":/resources/online-tile-providers.json")
     .WithMapProviders(":/resources/map-providers.json");
     .Init()
```

## Database access

I replaced database mutex with read-write lock, that will enable more readers at same moment. New functionality that is accessing database should not be added to `DBThread` (it should care just about databases, not for routing, searching...), but should be created `DBJob` that will acquire reader lock and do its business: 

```
  loadJob=new DBLoadJob(lookupProjection,maximumAreaLevel,/* lowZoomOptimization */ true);
  
  connect(loadJob, SIGNAL(databaseLoaded(QString,QList<osmscout::TileRef>)),
          this, SLOT(onDatabaseLoaded(QString,QList<osmscout::TileRef>)));
  connect(loadJob, SIGNAL(finished(QMap<QString,QMap<osmscout::TileId,osmscout::TileRef>>)),
          this, SLOT(onLoadJobFinished(QMap<QString,QMap<osmscout::TileId,osmscout::TileRef>>)));

  dbThread->RunJob(loadJob);
```

DBJob have to release its reader lock when its job is finished or when it is destroyed...

### CODE REVIEW

**Comment by Framstag** [comment]
I'M currently looking through the code (looks good up to now). The Job API you mentioned look at little bit like working with Promise/Future. Does Qt provide something similar and would it be helpful to use such pattern? You will likely still need the signal/slot mechanism for getting notified though, too?

File: libosmscout-client-qt/src/osmscout/DBJob.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    for (auto &tile:tiles){
+      if (tile->IsComplete()){
+        emit tileStateChanged(path,tile);
+      }
+    }
+
+  }
+}
+
+void DBLoadJob::onTileStateChagned(QString dbPath,const osmscout::TileRef tile)
```
Comment (Framstag) [inline libosmscout-client-qt/src/osmscout/DBJob.cpp]: Spelling mistake in function name

Comment (Framstag) [inline libosmscout-client-qt/src/osmscout/DBJob.cpp]: const QString& for dbPath to optimize? I also use const& for XXXRef classes, though that usage could be discussed.

**Comment by Karry** [comment]
Hi. Exactly. I expect that jobs may report computation progress (router, load...) or partial result (search, reverse lookup...) by signals. Futures dont fit to this.

With reference usages, I am not sure how it works with queued Qt signal connection. I expect that all arguments are copied anyway...? I will try to find it.

**Comment by aquiles2k** [comment]
Sorry for not responding earlier, this looks good so far, though i cant really say truly until i have to use it myself. As far as i understand this has not yet decoupled different view rendering?

**Comment by Karry** [comment]
@aquiles2k : Not yet, rendering is still in shared DBThread... I am working on it here: https://github.com/Karry/libosmscout/tree/qt-map-renderers plane rendering is rewritten, but there is some bug (probably concurrency) that cause random segfaults... help is welcome :-)

**Comment by aquiles2k** [comment]
Ah thats pretty far along - i could have saved me the initial look at the pipeline in #284 then :)

So it looks like your giving each MapWidget its own MapRenderer which handles the rendering stuff the dbthread did before? I'll have a look.

**Comment by aquiles2k** [comment]
Ok i tested it and it works great so far for me - no segfaults. Do you have some reproducible way i should test?

I tested it with just adding a second very simple Map below the other Map in OSMScout2 and started heavily paning and zooming in both maps.

**Comment by aquiles2k** [comment]
Ok right after writing the last comment i got a segfault :)

Interestingly enough it would the crash on every new startup with the first render. Unfortunately as soon as i started to put in debug logging, it disappeared. I did also found out that putting qDebug() calls in OSMScoutQtBuilder::Init will segfault it too :)

**Comment by aquiles2k** [comment]
Ok some pretty good way to let it segfault for me is to resize the whole Application window - probably because he then gets redraw calls for both views at nearly the same time.

**Comment by aquiles2k** [comment]
@Karry So after finding out that the resize crashes was a problem on my side i retried and so far couldnt get a crash. Do you have some reproducible way on your side for the segfaults?

**Comment by Karry** [comment]
It seems that segfaults was caused by using (read) `QList` instance from multiple threads. `std::list` is (should be) safe for using from multiple thread (for read). It seems to be stable now...

**Comment by Karry** [comment]
Hi. It seems that there is (was) several causes of segfaults. One was caused by using `QList` (read) instance with databases from multiple threads. `std::list` is (should be) safe for using from multiple threads (for read)...

Another segfault is probably caused by `DataFile` that has some methods without mutex synchronisation...

Lets move discussion here: https://github.com/Framstag/libosmscout/pull/336 It seems that there is a lot of fun with multithreading bugs :-(

**Pull request merged**